File: C++/2D-Arrays/IdentityMatrixCheck.cpp
/*
Approach:
1. To check if a matrix is an identity matrix, it needs to be a square matrix, i.e., the number of rows should be equal to the number of columns.
2. Loop through the main diagonal (i.e., elements where row index == column index) and check if they are equal to 1. If any diagonal element is not 1, the matrix is not an identity matrix.
3. Loop through the non-diagonal elements (i.e., elements where row index != column index) and check if they are equal to 0. If any non-diagonal element is not 0, the matrix is not an identity matrix.
4. If all conditions are met, the matrix is an identity matrix.

Time Complexity: The time complexity of this program is O(n^2), where 'n' is the size of the square matrix.

Space Complexity: The space complexity of this program is O(n^2) because we are using a 2D array to store the input matrix.

Sample Input:
Enter the size of the square matrix: 3
Enter the elements of the matrix:
1 0 0
0 1 0
0 0 1

Sample Output:
The given matrix is an Identity Matrix.

Sample Input:
Enter the size of the square matrix: 3
Enter the elements of the matrix:
1 0 0
0 2 0
0 0 1

Sample Output:
The given matrix is NOT an Identity Matrix.

In the first sample input, all the diagonal elements are 1, and all non-diagonal elements are 0, making it an identity matrix. In the second sample input, the element at mat[1][1] is 2, making it not an identity matrix.
*/

#include <iostream>
using namespace std;

bool isIdentityMatrix(int mat[][100], int n) {
    // Check for square matrix
    for (int i = 0; i < n; i++) {
        if (mat[i][i] != 1) {
            return false;
        }
    }

    // Check for non-diagonal elements
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            if (i != j && mat[i][j] != 0) {
                return false;
            }
        }
    }

    return true;
}

int main() {
    int n;
    cout << "Enter the size of the square matrix: ";
    cin >> n;

    int mat[100][100];
    cout << "Enter the elements of the matrix:" << endl;
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            cin >> mat[i][j];
        }
    }

    if (isIdentityMatrix(mat, n)) {
        cout << "The given matrix is an Identity Matrix." << endl;
    } else {
        cout << "The given matrix is NOT an Identity Matrix." << endl;
    }

    return 0;
}

File: Python/2D-Arrays/IdentityMatrixCheck.py
'''
Approach:
1. First, we need to check if the matrix is square (the number of rows is equal to the number of columns). If it's not square, it cannot be an identity matrix.
2. Next, we check if all the diagonal elements are 1 and all other elements are 0. If this condition is met, the matrix is an identity matrix.

Time Complexity: The time complexity of this program is O(n^2), where 'n' is the number of rows (or columns) in the matrix. This is because we need to traverse all the elements of the matrix once.

Space Complexity: The space complexity is O(1) because we are not using any additional data structures that grow with the input size. We are only using a few variables to keep track of the matrix dimensions and loop indices.

Sample Input:
sample_matrix = [
    [1, 0, 0],
    [0, 1, 0],
    [0, 0, 1]
]

Sample Output:
The given matrix is an identity matrix.
'''

def is_identity_matrix(matrix):
    rows = len(matrix)
    columns = len(matrix[0])

    # Check if the matrix is square
    if rows != columns:
        return False

    # Check if all diagonal elements are 1 and all other elements are 0
    for i in range(rows):
        for j in range(columns):
            if i == j and matrix[i][j] != 1:  # Check diagonal elements
                return False
            if i != j and matrix[i][j] != 0:  # Check non-diagonal elements
                return False

    return True

# Sample input
sample_matrix = [
    [1, 0, 0],
    [0, 1, 0],
    [0, 0, 1]
]

# Check if the sample matrix is an identity matrix
if is_identity_matrix(sample_matrix):
    print("The given matrix is an identity matrix.")
else:
    print("The given matrix is not an identity matrix.")

